### ISSUE TITLE
constants cannot be attributes of another class

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5797]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5797|
|Type|defect|
|Created|2012-07-20 12:41:16Z|
|Branch|https://github.com/twisted/twisted/tree/constants-foreign-attribute-5797|

This behavior is surprising and a bit unfortunate:

```python
>>> from twisted.python.constants import Names, NamedConstant
>>> 
>>> class Foo(Names):
...     x = NamedConstant()
... 
>>> class Bar(object):
...     a = Foo.x
... 
>>> Bar.a
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/constants.py", line 40, in __get__
    cls._initializeEnumerants()
AttributeError: type object 'Bar' has no attribute '_initializeEnumerants'
>>> Bar().a
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/constants.py", line 40, in __get__
    cls._initializeEnumerants()
AttributeError: type object 'Bar' has no attribute '_initializeEnumerants'
>>> 
```

People generally expect to be able to take an object and make it an attribute of a class like this.

Perhaps `NamedConstant` needs to check the type of the class it's being looked up on?  Alternatively, maybe the experiment in not using metaclasses is a failure and `twisted.python.constants` needs to be switched to using metaclasses.

<details><summary>Searchable metadata</summary>

```
trac-id__5797 5797
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_constants_foreign_attribute_5797 branches/constants-foreign-attribute-5797
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1342788076000000 1342788076000000
changetime__1362087579000000 1362087579000000
version__None None
owner__exarkun exarkun
cc__hynek
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [37050]) Branching to 'constants-foreign-attribute-5797'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/constants-foreign-attribute-5797)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_4"></a>|@tomprince set owner to @exarkun|
|-|-|

  1. The comment in `test_initializedOnce` is no longer true.
  1. There don't appear to be tests that other user-defined attributes aren't accessed by `lookupByName`.
  1. Is there any reason for support `_ConstantsContainerType` instances that don't have '_constantType`? In any case, it appears that, despite the docstring, it only supports ones that have `_constantType` of `None`.
     * I guess this is used for the base class `_ConstantsContainer`.
     * It occurs to me that most of the behavior (except `__new__`) could probably be implemented as instance methods on `_ConstainsContainerType` instead of class-methods on `_ContainerClass`.
  1. What is going to happen if somebody tries to assign a constant as an additional attribute on a different constant-container of the same type. I believe that things will silent break. (Please file a ticket about dealing with this case)
  1. There don't appear to be any tests about having additional attributes on constant-containers, although there is code to support that case. (Please file a ticket for adding the relevant tests).

Please resubmit for review after addressing the above issues.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner|
|-|-|

Thanks for the review.

> The comment in test_initializedOnce is no longer true.

Fixed in r37313 and r37314.

> There don't appear to be tests that other user-defined attributes aren't accessed by lookupByName.

`test_notLookupMissingByName` covers this.  I wonder what additional coverage you're looking for.

> Is there any reason for support _ConstantsContainerType instances that don't have '_constantType? In any case, it appears that, despite the docstring, it only supports ones that have _constantType of None`.

Fixed in r37315.

> I guess this is used for the base class _ConstantsContainer.

Yes.

> It occurs to me that most of the behavior (except __new__) could probably be implemented as instance methods on _ConstainsContainerType instead of class-methods on _ContainerClass.

Probably.  I'd rather not bother changing it, though.  It works as-is.

> What is going to happen if somebody tries to assign a constant as an additional attribute on a different constant-container of the same type. I believe that things will silent break. (Please file a ticket about dealing with this case)

This is a somewhat bad behavior, I didn't want to introduce this change without fixing it.  Fixed it in r37317.

> There don't appear to be any tests about having additional attributes on constant-containers, although there is code to support that case. (Please file a ticket for adding the relevant tests).

If I comment out the `isinstance(descriptor, cls._constantType)` check in `__new__`, the test module fails even to import.  is that the support code you're talking about, or is there something else?

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_6"></a>|@tomprince set owner to @exarkun|
|-|-|

The changes looks good. Please merge.

>> There don't appear to be tests that other user-defined attributes aren't accessed by lookupByName.
> test_notLookupMissingByName covers this. I wonder what additional coverage you're looking for. 
>> There don't appear to be any tests about having additional attributes on constant-containers, although there is code to support that case. (Please file a ticket for adding the relevant tests).
> If I comment out the isinstance(descriptor, cls._constantType) check in __new__, the test module fails even to import. is that the support code you're talking about, or is there something else?

None of the `_ConstantContainer` subclasses in the tests have any explicit attributes other than constants. Something like
```
!python
class Test(Names):
    bar = "Some text"
    @classmethod
    def baz(cls, arg):
       doStuff()
```
but doing that is explicitly suggested in the howto, There should be some explicit tests that those work as expected.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [37320]) Give METHOD another random attribute and verify it is not looked up by lookupByName

refs [#5797](https://github.com/twisted/twisted/issues/5797)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_9"></a>|@tomprince set owner to @exarkun|
|-|-|

> The changes looks good. Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [37325]) Merge constants-foreign-attribute-5797

Author: exarkun
Reviewer: tom.prince
Fixes: [#5797](https://github.com/twisted/twisted/issues/5797)

Allow constant values defined using `twisted.python.constants` to be set as attributes of other
classes and then be retrieved again via attribute access.
